---
title: "group_03_package"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{group_03_package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(Lab8Package)
```

The package "group 3" is a simple DNA/RNA sequence tool. 
Link to our GitHub repository: https://github.com/rforbiodatascience25/group_03_package

This package provides a small collection of functions for generating, manipulating, and analyzing nucleotide and amino acid sequences. It allows users to create random DNA sequences, transcribe DNA to RNA, split sequences into codons, translate codons into amino acid sequences, and visualize amino acid composition as bar plots.


Everyone writes about their function and some code for explanation 


about and demonstrate how each function in the package is used (individually and in conjunction with each other). 
 
As a final section, discuss use cases for the package and what other functions could be included. Also include the main points from your discussion in Task 3.



Function 2:

This function takes a DNA sequence in argument and returns the corresponding RNA sequence. The process that is realized is changing the "T" into "U".
For instance, the "ATTGCT" DNA sequence will be converted into "AUUGCU".

```{r}
convert_T_U("ATTGCT")
```

In our process, we will use the random DNA sequence created with function 1 and make it an RNA one.

Function 3:

This function is named read_codons() and it takes in a DNA sequence as a string and outputs a character vector containing the codons in the sequence. It optionally takes in a "start" argument, which is 1 by default and indicates the position in the sequence you want to start reading codons from. 

````{r}
sequence <-  "AUUGCU"
read_codons(sequence)
other_sequence <- "AAUUGCU"
read_codons(sequence, 2)
````

Function 4:

This function "codons_into_aa" uses the data table we integrated. The "data_codon.rda" table links the trios of codons to the corresponding amino-acid.
For instance, the codon ("UUU", "GAC", "AAG") will become "FDK" after going through the function.

```{r}
codons <- c("UUU", "GAC", "AAG")
codons_into_aa(codons)
```
In our process, we use the sequence randomly generated that previously had been groupped in trios with function 3 and we translate it into amino-acids.
